File: scala/0072-edit-distance.scala
object Solution {
    def minDistance(word1: String, word2: String): Int = {
        var minDisArr = Array.ofDim[Int](word1.size+2, word2.size+2)
        for(i<- 0 to word1.size; j<- 0 to word2.size){
            if(i == 0 || j == 0) minDisArr(i)(j) = i + j

            else if (word1(i-1).equals(word2(j-1)))
                minDisArr(i)(j) = minDisArr(i-1)(j-1)
            
            else 
                minDisArr(i)(j) = 1 + (minDisArr(i - 1)(j - 1) min minDisArr(i)(j - 1) min minDisArr(i - 1)(j))

        }
        minDisArr(word1.size)(word2.size)
    }
}

File: rust/0072-edit-distance.rs
impl Solution {
    pub fn min_distance(word1: String, word2: String) -> i32 {
        use std::cmp::min;
        let (m, n) = (word1.len(), word2.len());
        let w1: Vec<char> = word1.chars().collect();
        let w2: Vec<char> = word2.chars().collect();
        let mut dp: Vec<usize> = (0..m + 1).collect();
        let mut temp;
        for i in 0..n {
            temp = vec![0; m + 1];
            temp[0] = i + 1;
            for j in 0..m {
                temp[j + 1] = if w1[j] == w2[i] {
                    dp[j]
                } else {
                    min(temp[j], min(dp[j + 1], dp[j])) + 1
                };
            }
            dp = temp;
        }
        dp[dp.len() - 1] as i32
    }
}
